src/gui/torrentcontentmodel.cpp

```diff
@@ -497,33 +530,25 @@ void TorrentContentModel::setupModelData(const BitTorrent::AbstractFileStorage &
     qDebug("Torrent contains %d files", filesCount);
     m_filesIndex.reserve(filesCount);
 
-    TorrentContentModelFolder *currentParent;
+    FolderAdaptor::FolderAdaptorPtr rootFolderAdaptor = std::make_shared<FolderAdaptor>(m_rootItem);
+    FolderAdaptor::FolderAdaptorPtr currentParent = nullptr;
     // Iterate over files
     for (int i = 0; i < filesCount; ++i)
     {
-        currentParent = m_rootItem;
+        currentParent = rootFolderAdaptor;
         const QString path = Utils::Fs::toUniformPath(info.filePath(i));
 
         // Iterate of parts of the path to create necessary folders
         QList<QStringView> pathFolders = QStringView(path).split(u'/', Qt::SkipEmptyParts);
         pathFolders.removeLast();
 
-        for (const QStringView pathPart : asConst(pathFolders))
-        {
-            const QString folderPath = pathPart.toString();
-            TorrentContentModelFolder *newParent = currentParent->childFolderWithName(folderPath);
-            if (!newParent)
-            {
-                newParent = new TorrentContentModelFolder(folderPath, currentParent);
-                currentParent->appendChild(newParent);
-            }
-            currentParent = newParent;
-        }
+        for (const QStringView folderName : asConst(pathFolders))
+            currentParent = currentParent->findChildFolderOrInsert(folderName.toString());
+
         // Actually create the file
-        TorrentContentModelFile *fileItem = new TorrentContentModelFile(
-                    Utils::Fs::fileName(info.filePath(i)), info.fileSize(i), currentParent, i);
-        currentParent->appendChild(fileItem);
-        m_filesIndex.push_back(fileItem);
+        TorrentContentModelFile *file = currentParent->insertFile(
```

Comment:
`const TorrentContentModelFile *file`

Reply:
But then you can't add it to `QVector<TorrentContentModelFile *>`.